fuzztest/core/client_channel/retry_state_test.cc

```diff
@@ -0,0 +1,389 @@
+    .WithDomains(OptionalOf(AnyRetryMethodConfig()), AnyServerThrottleData());
+
+void NoPolicyNeverRetries(std::vector<ServerMetadataHandle> md,
+                          bool committed_at_end) {
+  RetryState retry_state(nullptr, nullptr);
+  for (auto it = md.begin(); it != md.end(); ++it) {
+    EXPECT_EQ(
+        retry_state.ShouldRetry(**it, (it + 1 == md.end() && committed_at_end),
+                                FuzzerDebugTag),
+        absl::nullopt);
+  }
+}
+FUZZ_TEST(MyTestSuite, NoPolicyNeverRetries)
+    .WithDomains(VectorOf(AnyServerMetadata()), Arbitrary<bool>());
+
+void SuccessfulRequestsNeverRetry(
+    internal::RetryMethodConfig policy, ServerMetadataHandle md, bool committed,
+    RefCountedPtr<internal::ServerRetryThrottleData> throttle_data) {
+  RetryState retry_state(&policy, throttle_data);
+  EXPECT_EQ(retry_state.ShouldRetry(*md, committed, FuzzerDebugTag),
+            absl::nullopt);
+}
+FUZZ_TEST(MyTestSuite, SuccessfulRequestsNeverRetry)
+    .WithDomains(AnyRetryMethodConfig(), AnySuccessfulMetadata(),
+                 Arbitrary<bool>(), AnyServerThrottleData());
+
+void CommittedRequestsNeverRetry(
+    internal::RetryMethodConfig policy, ServerMetadataHandle md,
+    RefCountedPtr<internal::ServerRetryThrottleData> throttle_data) {
+  RetryState retry_state(&policy, throttle_data);
+  EXPECT_EQ(retry_state.ShouldRetry(*md, true, FuzzerDebugTag), absl::nullopt);
+}
+FUZZ_TEST(MyTestSuite, CommittedRequestsNeverRetry)
+    .WithDomains(AnyRetryMethodConfig(), AnyServerMetadata(),
+                 AnyServerThrottleData());
+
+void NonRetryableRequestsNeverRetry(
+    internal::RetryMethodConfig policy, ServerMetadataHandle md, bool committed,
+    RefCountedPtr<internal::ServerRetryThrottleData> throttle_data) {
+  RetryState retry_state(&policy, throttle_data);
+  EXPECT_EQ(retry_state.ShouldRetry(*md, committed, FuzzerDebugTag),
+            absl::nullopt);
+}
+FUZZ_TEST(MyTestSuite, NonRetryableRequestsNeverRetry)
+    .WithDomains(
+        RetryMethodConfigWithRetryableStatusCodes({GRPC_STATUS_ABORTED}),
+        ServerMetadataWithStatus(AnyStatusExcept(GRPC_STATUS_ABORTED)),
+        Arbitrary<bool>(), AnyServerThrottleData());
+
+void NeverExceedMaxAttempts(
+    internal::RetryMethodConfig policy, std::vector<ServerMetadataHandle> md,
+    bool committed_at_end,
+    RefCountedPtr<internal::ServerRetryThrottleData> throttle_data) {
+  RetryState retry_state(&policy, nullptr);
+  int attempts_completed = 0;
+  for (auto it = md.begin(); it != md.end(); ++it) {
+    ++attempts_completed;
+    if (retry_state.ShouldRetry(**it, (it + 1 == md.end() && committed_at_end),
+                                FuzzerDebugTag) == absl::nullopt) {
+      break;
+    }
+  }
+  EXPECT_LE(attempts_completed, policy.max_attempts());
+}
+FUZZ_TEST(MyTestSuite, NeverExceedMaxAttempts)
+    .WithDomains(AnyRetryMethodConfig(),
+                 VectorOf(AnyServerMetadata()).WithMaxSize(7),
```

Comment:
Not that it really matters here, but just FYI, we support a max of 5 configurable retries. :)

Reply:
I know :)

I wanted to overflow it sometimes, just in case.